File: sportsSchedule_algo.js
numOfTeams = 6

function createSchedule(numOfTeams) { //EACH TEAM SHOULD EVERYONE (HOME && AWAY) TOTAL#OFMATCHES = numOfTeams^2-numOfTeams
    allMatches = []
    for (var home = 1; home <= numOfTeams; home++) {
        for (var away = 1; away <= numOfTeams; away++) {
            if (home != away) {
                allMatches.push([home, away])
                // schedule.push([away, home])
            }
        }
    }
    matchSetupCount = 0
    matchesToPlay = Math.pow(numOfTeams, 2) - numOfTeams
    matchesPerWeek = numOfTeams / 2
    weeks = matchesToPlay / matchesPerWeek
    console.log(matchSetupCount)
    console.log(matchesToPlay)
    console.log(matchesPerWeek)
    console.log(weeks)

    for (var wk = 1; wk <= weeks; wk++) {
        allTeam = [] //EACH WEEK A MATCH INCLUDE EVERY TEAM MUST BE PLACED, BUT EACH TEAM MUST ONLY PLAY 1 MATCH A WEEK
        for (i = 1; i <= numOfTeams; i++) {
            allTeam.push(i)
        }
        matchesConfirmed = 0
        matchAssigned = false
        // for (var match = 0; match < matchesPerWeek; match++) { //WITH EVERY TEAM PLAYING A MATCH: ONLY (numOfTeams/2) MATCHES WILL BE ASSIGNED PER WEEK 
        for (var allMatchCycle = 0; allMatchCycle < matchesToPlay; allMatchCycle++) {
            teamIndexes = [];
            matchSetupCount = 0
            if (allMatches[allMatchCycle].length < 3) {
                for (var verifyIndex = 0, availTeams = allTeam.length; verifyIndex < availTeams; verifyIndex++) {
                    //HOW TO PREVENT REPEATING WEEKS FOR SAME TEAM!!
                    //5,2 AND 5,3
                    //2,4 AND 2,6
                    if (allMatches[allMatchCycle][0] == allTeam[verifyIndex] || allMatches[allMatchCycle][1] == allTeam[verifyIndex]) {
                        teamIndexes.push(verifyIndex) //AS THE AVAILABLE TEAMS ARE FOUND, THEY WILL BE REMOVED FROM THE LIST OF TEAMS NEEDING TO BE ASSIGNED TO A MATCH FOR THE WEEK
                        matchSetupCount++
                        if (matchSetupCount == 2) { //ONCE WE DETERMINE BOTH TEAMS OF A MATCH ARE AVAILABLE TO PLAY, THEN WE WILL REMOVE THEM FROM THE AVAILABLE LIST AND FINALLY ASSIGN THE MATCH A WEEK TO BE PLAYED
                            for (var teamToRemove = 1; teamToRemove >= 0; teamToRemove--) {
                                // console.log('WEEK: ', wk,'match ', allMatches[allMatchCycle], 'availableTeam ', allTeam, 'teamIndex ', teamIndexes, 'matchSetUp ', matchSetupCount, 'allTeamLEN: ', allTeam.length)
                                if (allTeam.length > 1) {
                                    for (var allTeamRemoval = teamIndexes[teamToRemove], x = allTeam.length - 1; allTeamRemoval < x; allTeamRemoval++) {
                                        temp = allTeam[allTeamRemoval]
                                        allTeam[allTeamRemoval] = allTeam[allTeamRemoval + 1]
                                        allTeam[allTeamRemoval + 1] = temp
                                    }
                                }
                                allTeam.pop()
                            }
                            allMatches[allMatchCycle].push(wk) //ADDING THE WEEK THIS MATCH IS PLAYED TO THE allMatches
                            matchesConfirmed++
                            matchAssigned = true
                        }
                        if (matchAssigned == true) {
                            matchAssigned = false
                            break;
                        }
                    }
                }
            }
            if (matchesConfirmed == matchesPerWeek) {
                break;
            }
        }
        // }
    }

    for (i = 0; i < allMatches.length; i++) {
        if (allMatches[i].length == 3) {
            allMatches[i].push(Math.random())
        }
    }

    orderedLeague = []

    for (var byWeek = 1; byWeek <= weeks; byWeek++) {
        for (i = 0; i < allMatches.length; i++) {
            if (allMatches[i].length == 4) {
                if (allMatches[i][2] == byWeek) {
                    orderedLeague.push(allMatches[i])
                }
            }
        }
    }
    return orderedLeague
}
array = createSchedule(numOfTeams)
// console.log(leagueSchedule)

// array = [5, 6, 3, 1, 8, 7, 2, 4]
// randomOrder = []

// for (i = 0; i < array.length; i++) {
//     console.log(array[i])
//     randomOrder.push([array[i], Math.random()])
// }

for (x = 0; x < array.length; x++) {
    for (i = 0; i < array.length - 1; i++) {
        if (array[i][3] < array[i + 1][3]) {
            temp = array[i]
            array[i] = array[i + 1]
            array[i + 1] = temp
        }
    }
}

for (x = 0; x < array.length; x++) {
    array[x].pop()
}
console.log(array)

File: sportsSchedule_algo.py


def generateSchedule(numOfTeams):
    scheduleGrid = []
    weeklyGrid = []
    for x in range(numOfTeams):  # Create a grid for all matches to be played
        scheduleGrid.append([])
        weeklyGrid.append([])

    allMatches = []
    for home in range(1, numOfTeams+1):  # Insert all matches to be played
        for away in range(1, numOfTeams+1):
            allMatches.append([home, away])

    match = 0
    team = 1
    for x in range(len(scheduleGrid)):
        while match < len(allMatches):
            if allMatches[match][0] != team and scheduleGrid[x][len(scheduleGrid) - 1] != None:
                break

            if allMatches[match][0] != allMatches[match][1]:
                scheduleGrid[x].append(allMatches[match])
                weeklyGrid[x].append([])

            else:
                scheduleGrid[x].append(None)
                weeklyGrid[x].append([])

            match += 1

        team += 1

    matchesToPlay = pow(numOfTeams, 2) - numOfTeams
    matchesPerWeek = int(numOfTeams / 2)
    weeks = int(matchesToPlay / matchesPerWeek)

    for nullValue in range(numOfTeams):
        weeklyGrid[nullValue][nullValue].append(None)


    x = 0
    for weekCount in range(1, int(weeks)+1):
        for row in range(len(weeklyGrid)):
            for col in range(len(weeklyGrid[row])):
                if row == col:
                    continue
                rowTotal = 0
                for tempCol in range(numOfTeams):
                    if len(weeklyGrid[row][tempCol]) != 0 and weeklyGrid[row][tempCol][0] != None:
                        rowTotal += 1
                colTotal = 0
                for tempRow in range(numOfTeams):
                    if len(weeklyGrid[tempRow][col]) != 0 and weeklyGrid[tempRow][col][0] != None:
                        colTotal += 1
                if rowTotal == weekCount:
                    break
                if colTotal == weekCount:
                    continue
                if row > col:
                    continue
                if len(weeklyGrid[row][col]) == 0:
                    weeklyGrid[row][col].append(weekCount)
                    weeklyGrid[col][row].append(
                        int(weekCount + ((pow(numOfTeams, 2) - numOfTeams) / (numOfTeams / 2) / 2)))
                    break

    for homeTeam in range(numOfTeams):
        for awayTeam in range(numOfTeams):
            if scheduleGrid[homeTeam][awayTeam] == None:
                continue
            scheduleGrid[homeTeam][awayTeam].append(weeklyGrid[homeTeam][awayTeam][0])

    leagueSchedule = []
    for weekCount in range(1, weeks+1):
        matchesConfirmed = 0
        for homeTeam in range(numOfTeams): #Row
            for awayTeam in range(numOfTeams): #Col
                if scheduleGrid[homeTeam][awayTeam] != None:
                    if scheduleGrid[homeTeam][awayTeam][2] == weekCount:
                        leagueSchedule.append(scheduleGrid[homeTeam][awayTeam])
                        matchesConfirmed += 1
                        #if scheduleGrid[homeTeam][awayTeam][0] or [1] == homeTeam break awayTeam FOR
                        if scheduleGrid[homeTeam][awayTeam][0] == homeTeam+1 or scheduleGrid[homeTeam][awayTeam][1] == homeTeam+1:
                            break
                if matchesConfirmed == matchesPerWeek:
                    break
            if matchesConfirmed == matchesPerWeek:
                break
    return leagueSchedule

numOfTeams = 16
Schedule  = generateSchedule(numOfTeams)
print(len(Schedule))